sdp/pyfftw_sdp.py

```diff
@@ -4,48 +4,99 @@
 
 class SLIDING_DOT_PRODUCT:
     # https://stackoverflow.com/a/30615425/2955541
-    def __init__(self):
-        self.m = 0
+    def __init__(self, max_n=2**20):
+        """
+        Parameters
+        ----------
+        max_n : int
+            Maximum length to preallocate arrays for. This will be the size of the
+            the real-valued array. A complex-valued array of size `1 + (max_n // 2)`
+            will also be preallocated.
+        """
         self.n = 0
-        self.threads = 1
-        self.rfft_Q_obj = None
-        self.rfft_T_obj = None
-        self.irfft_obj = None
-
-    def __call__(self, Q, T):
-        if Q.shape[0] != self.m or T.shape[0] != self.n:
-            self.m = Q.shape[0]
+        self.next_fast_n = 0
+
+        # Preallocate arrays
+        self.real_arr = pyfftw.empty_aligned(max_n, dtype="float64")
+        self.complex_arr = pyfftw.empty_aligned(1 + (max_n // 2), dtype="complex128")
+
+        # Store FFTW objects in a dict keyed by `next_fast_n`, where n=len(T)
```

Comment:
@seanlaw 
Note that, in `irfft`, the length of OUTPUT is `next_fast_n` (NOT the length of INPUT). However, since computing `SDP(Q, T)` requires both `rfft` and `irfft` objects, and `next_fast_n` is the length based on which the length of input/output arrays are determined, I am using the key `next_fast_n` for both objects.

Reply:
> Should the dictionary key be the tuple (next_fast_n, threads, planning_flag) rather than just next_fast_n?

1. I suggest that we use `n_threads` instead of `threads`
2. Is it actually possible to use `planning_flag` as a part of the tuple? If so, having all three in a tuple is fine

Reply:
> I suggest that we use `n_threads` instead of `threads`

It makes sense. Will do.

> Is it actually possible to use `planning_flag` as a part of the tuple? 

The value provided for `planning_flag` has the `string` type. It should be okay to use it as part of the tuple. I will improve docstring to explain that the following values are supported for `planning_flag`: 
*  `FFTW_ESTIMATE`
* `FFTW_MEASURE`
* `FFTW_PATIENT` 
* `FFTW_EXHAUSTIVE`